lib/recursive-methods.rb

```diff
@@ -1,49 +1,108 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: also O(n), where a new method is stored n times in the call stack
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  end
+  if n <= 1
+    return 1
+  else
+    n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: The method is called n/2 times, and removing the constant we get O(n)
+# Space complexity: Since the method creates a new array every time it is called and stores a method in the call stack, it takes up 2n memory, but removing the constant makes the space complexity O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length > 1
+    first_char = s[0]
+    last_char = s[-1]
+    s = reverse(s.delete_prefix(s[0]).delete_suffix(s[-1]))
```

Comment:
`delete_prefix` & `delete_suffix`  create new arrays and  copy all the individual elements over and so this line is O(n) by itself.